apps/volunteer/sign_up.py

```diff
@@ -136,3 +132,52 @@ def account():
     form.allow_comms.data = volunteer.allow_comms_during_event
 
     return render_template("volunteer/account.html", user=current_user, form=form)
+
+
+def generate_arrival_options():
+    choices = []
+
+    # Work out our first arrival based on config
+    first_arrival = event_start() - timedelta(app.config["ARRIVAL_DAYS"])
+
+    # Work out dates between first arrival and end of the event
+    choices = generate_day_options(first_arrival, event_end())
+
+    # Replace first array element with first date and 'or earlier'
+    choices[0] = (
+        first_arrival.strftime("%F"),
+        first_arrival.strftime("%A %-d %B or earlier"),
+    )
+
+    return choices
+
+
+def generate_departure_options():
+    choices = []
+
+    # Work out our last arrival based on config
+    last_departure = event_end() + timedelta(app.config["DEPARTURE_DAYS"])
+
+    # Work out dates between start of the event and last departure
+    choices = generate_day_options(event_start(), last_departure)
+
+    # Replace last array element with the last date and 'or later'
+    choices[len(choices) - 1] = (
+        last_departure.strftime("%F"),
+        last_departure.strftime("%A %-d %B or later"),
+    )
+
+    return choices
```

Comment:
If and when `generate_day_options` is refactored as suggested, then you could collapse this down to:

```
def generate_departure_options():
    end = event_end() + timedelta(app.config["DEPARTURE_DAYS"])
    day_before_end = end - timedelta(days=1)

    for date in period(event_start(), day_before_end):
        yield generate_day_option(date)
    yield generate_day_option(end, ' or later')
```